### PR TITLE
Feature/pattern matching

### LINKED ISSUE #1954: Pattern Matching
Would there be interest in adding a pattern matching macro to contrib, or would that have to go in an external library?

I have a working proof of concept right now, and it seems pretty similar to destructuring #1940, so I thought they might go well together.

### PR DESCRIPTION
**Pull request by allison-casey**
closes #1954 

adds `match` special form for > 3.10. This is a pretty literal translation of the python syntax so it should be easy to learn the syntax. 

had to add `end_line` & `end_column` to `Asty` and `Object` since they are required on some new nodes, this doesn't affect other nodes. seems like it makes that info more consistent with `start_line` and `start_column` anyways. 

This implementation is heavily parser based since its heavily recursive and would have been a nightmare to do by hand. As such the syntax errors are a bit obtuse, but we already knew that https://github.com/vlasovskikh/funcparserlib/issues/52

This solves the issue of arbitraty statements in the `:if` guard by lifting them up into a function. this works because all of the bind targets have already been bound into local scope by the time the function gets called.

```clojure
=> (match [0 1 2]
...     (case [0 #* x] :as z :if (as-> z $ (+ $ [3]) (len $) (= $ 4))
...         0))
def _hy_anon_var_2():
    hyx_Xdollar_signX = z
    hyx_Xdollar_signX = hyx_Xdollar_signX + [3]
    hyx_Xdollar_signX = len(hyx_Xdollar_signX)
    hyx_Xdollar_signX = hyx_Xdollar_signX == 4
    return hyx_Xdollar_signX
match [0, 1, 2]:
    case [0, *x] as z if _hy_anon_var_2():
        _hy_anon_var_1 = 0
    case _:
        _hy_anon_var_1 = None
_hy_anon_var_1
0

```

let me know if you manage to blow it up somehow :+1:

### CODE REVIEW

**Comment by allison-casey** [comment]
I'm realizing now that this will break when using let since the match values aren't protected so they'll get replaced

**Comment by Kodiologist** [comment]
`let` is truly the technical debt that keeps on giving.

**Comment by Kodiologist** [comment]
I think the head symbol `case` isn't necessary. Every direct child form of `match` already has to be a case clause. So the syntax can be simplified to something like:

```
(match (Point -1 1)
  [(Point 0 0) "The Origin"]
  [(Point 0 :y y) f"One on the Y axis at {y}"]
  [[(Point :x 0 y1) (Point 0 :y y2)] f"Two on the Y axis at {y1}, {y2}"]
  [_ "some place off in the ether"])
```

**Comment by Kodiologist** [comment]
One could even imagine an unbracketed syntax, as I know you're thinking of changing `cond`:

```
(match (Point -1 1)
  (Point 0 0) "The Origin"
  (Point 0 :y y) f"One on the Y axis at {y}"
  [(Point :x 0 y1) (Point 0 :y y2)] f"Two on the Y axis at {y1}, {y2}"
  _ "some place off in the ether")
```

In this design, only one body form is allowed (possibly preceded by `:as` and `:if` parts), so you'd use `do` if you needed more than one.

File: hy/compiler.py [hunk 1]
```diff
@@ -598,6 +621,166 @@ def compile_raise_expression(self, expr, root, exc, cause):
...
+        _match_value + _assignment
+        | brackets(many(_pattern | unpack("iterable"))) + _assignment
+        | pexpr(keepsym("|"), many(_pattern)) + _assignment
+        | pexpr(keepsym(","), many(_pattern)) + _assignment
+        | curly_brackets(many(LITERAL + _pattern), maybe(pvalue("unpack-mapping", SYM))) + _assignment
+        | pexpr(
+            notsym("|", ",", "unpack-mapping", "unpack-iterable"),
+            many(_pattern),
+            many(KEYWORD + _pattern),
+        ) + _assignment
```
Comment (scauligi) [inline hy/compiler.py:641]: The ` + _assignment` can be factored out if you stick the alternations inside a set of parentheses, no?

File: hy/compiler.py [hunk 2]
```diff
@@ -598,6 +621,166 @@ def compile_raise_expression(self, expr, root, exc, cause):
...
+    )
+    match_clause = _pattern + maybe(sym(":if") + FORM) + _assignment
+
+    @special(
+        (PY3_10, "match"), [FORM, oneplus(pexpr(sym("case"), match_clause, many(FORM)))]
```
Comment (scauligi) [inline hy/compiler.py:646]: Agreed with @Kodiologist we should drop the "case"; and a mild preference for the unbracketed version of the syntax. Also this should be `many` instead of `oneplus` to match other parts of Hy that allow vacuous constructs.

File: hy/compiler.py [hunk 3]
```diff
@@ -598,6 +621,166 @@ def compile_raise_expression(self, expr, root, exc, cause):
...
+
+            if bad_assignment:
+                raise self._syntax_error(bad_assignment, ":as assignment form cannot come after an :if guard")
+
+            body = self._compile_branch(body)
+            body += asty.Assign(pattern[0], targets=[return_var], value=body.force_expr)
+            body += body.expr_as_stmt()
+            body = body.stmts
+
+            if pattern[0] == Symbol("_"):
```
Comment (scauligi) [inline hy/compiler.py:667]: `case catchall: ...` where `catchall` is a valid symbol is allowed in match statements to name the catchall case, so this should be
```python
```

**Review by scauligi — COMMENTED** [review]
Yeah I agree with @Kodiologist on dropping the "case" root. I don't have a strong preference on bracketed vs unbracketed but I'm leaning slightly towards unbracketed.

I can take a look at how this interacts with `let`, it might not be so bad (fingers crossed)

File: hy/compiler.py (hunk 1 above)
Comment (allison-casey) [reply hy/compiler.py]: good catch! done

File: hy/compiler.py (hunk 2 above)
Comment (allison-casey) [reply hy/compiler.py]: valid, done.

File: hy/compiler.py (hunk 3 above)
Comment (allison-casey) [reply hy/compiler.py]: fixed

**Comment by allison-casey** [comment]
i added a commit with bracketless match cases. give it a try and see how it feels. If we like it i'll squash it. 
if you wanna try the bracketed form all you need to do is change 
```python
[FORM, oneplus(pexpr(sym("case"), match_clause, many(FORM)))]
```
to 
```python
[FORM, oneplus(brackets(match_clause, many(FORM)))]
```

**Comment by scauligi** [comment]
Hmm, thinking on this some more, I don't think we should result in `None` for an implicit default case, it feels too much like silent failure.

We should either explicitly raise an Exception of some sort in the implicit case (similar to some other match language implementations I've seen) or just not include an implicit default at all (which reflects the equivalent Python more closely).

In the latter case, if the user tries to use/access the result of an unmatched `match`, they'll get a `NameError: name '_hy_anon_var_#' is not defined`, but the traceback still correctly points to the match form. If they don't use the result (e.g. using a partial `match` just for some side effects) then no exceptions will get thrown.

**Comment by scauligi** [comment]
Sorry to step on your toes @allison-casey but I kind of sniped myself on the `let` fix. By all means edit/rearrange as you see fit.

Here's the patch for `let` (presented without tests sorry):
```diff
diff --git a/hy/contrib/walk.hy b/hy/contrib/walk.hy
index e683b82c..65e2cf31 100644
--- a/hy/contrib/walk.hy
+++ b/hy/contrib/walk.hy
@@ -320,6 +320,41 @@
                                             (if (and tail (= (len (get tail 0)) 2))
                                                 #{(get tail 0 0)}
                                                 #{})))))
+
+  (defn handle-match [self]
+    ;; protect name bindings from match patterns
+    (setv [expr #* cases] (self.tail)
+          new-expr (self.expand-symbols expr)
+          new-cases [])
+    (for [[pattern body] (by2s cases)]
+      ;; collect protected names from pattern
+      ;; before evaluating the corresponding body
+      (setv protected #{})
+      (defn handle-match-symbol [form]
+        (if (in "." form)
+          (self.expand-symbols form)
+          (do
+            (.add protected form)
+            form)))
+      (defn handle-match-call [form]
+        (setv head (get form 0))
+        (setv tail (cut form 1 None))
+        (cond
+          [(= head '.) (self.expand-symbols form)]
+          [True `(~head ~@(traverse-pattern tail))]))
+      (defn handle-pattern-form [form]
+        (cond
+          [(symbol? form) (handle-match-symbol form)]
+          [(call? form) (handle-match-call form)]
+          [(coll? form) (traverse-pattern form)]
+          [True form]))
+      (defn traverse-pattern [pattern]
+        (walk handle-pattern-form (fn [x] x) pattern))
+      (setv new-pattern (handle-pattern-form pattern))
+      (setv new-body (self.expand-symbols body (| protected self.protected)))
+      (.extend new-cases [new-pattern new-body]))
+    `(~(self.head) ~new-expr ~@new-cases))
+
   (defn handle-args-list [self]
     (setv protected #{}
           argslist [])
@@ -407,6 +442,7 @@
       [(= head 'except) (self.handle-except)]
       [(= head ".") (self.handle-dot)]
       [(= head 'defclass) (self.handle-defclass)]
+      [(= head 'match) (self.handle-match)]
       [(= head 'quasiquote) (self.+quote)]
         ;; must be checked last!
       [(in (hy.mangle head) _mangled-special) (self.handle-special-form)]
```

And also here's a quick patch for allowing `(. SYM...)` forms in match patterns to mirror `SYM.SYM...` forms:
```diff
diff --git a/hy/compiler.py b/hy/compiler.py
index 5b34b304..9f85bf93 100755
--- a/hy/compiler.py
+++ b/hy/compiler.py
@@ -632,13 +632,14 @@ class HyASTCompiler(object):
         (
             _match_value
             | brackets(many(_pattern | unpack("iterable")))
+            | pexpr(keepsym("."), many(SYM))
             | pexpr(keepsym("|"), many(_pattern))
             | pexpr(keepsym(","), many(_pattern))
             | curly_brackets(
                 many(LITERAL + _pattern), maybe(pvalue("unpack-mapping", SYM))
             )
             | pexpr(
-                notsym("|", ",", "unpack-mapping", "unpack-iterable"),
+                notsym(".", "|", ",", "unpack-mapping", "unpack-iterable"),
                 many(_pattern),
                 many(KEYWORD + _pattern),
             )
@@ -735,6 +736,13 @@ class HyASTCompiler(object):
                 value,
                 value=self.compile(value).expr,
             )
+        elif isinstance(value, Expression) and value[0] == Symbol("."):
+            root, syms = value
+            dotform = mkexpr(root, *syms).replace(value)
+            return asty.MatchValue(
+                value,
+                value=self.compile(dotform).expr,
+            )
         elif isinstance(value, Symbol):
             return asty.MatchAs(
                 value, name=mangle(value)
```

**Comment by Kodiologist** [comment]
> Hmm, thinking on this some more, I don't think we should result in `None` for an implicit default case, it feels too much like silent failure.

I disagree. Changing this would be a big departure from the rest of the language.

**Comment by allison-casey** [comment]
a failure to match anything in a `match` statement results in a no-op so so we kind of have to return None in the context of an s-expression

**Comment by allison-casey** [comment]
I pushed a commit with another potential syntax. One i'm actually quite liking having messed around with it for a bit. It replaces the `:if + FORM` syntax with wrapping the pattern in a `where` expression. This is pretty much the same as `fennel` match and i think results in a pretty clean looking expression

```clojure
(match [1 2 3 4]
  [1 _ x] :as z [x z]
  (where [_ _ x _] :as z (odd? x)) z
  (where z (= (len z) 4)) z)
```

**Comment by Kodiologist** [comment]
I don't see the advantage over the more Python-like `:if`. It's weird that you add `where` in order to add a condition to the test and yet the condition form comes after the pattern.

**Comment by allison-casey** [comment]
personally looks cleaner in my mind, but you're right that its not sufficiently better than python's syntax to merit creating the disparity. I can make it a macro easily anyways. Will delete it and squash bracketless

**Comment by allison-casey** [comment]
implemented let binding protection + fixed `_` incorrectly being bound to a symbol

**Comment by Kodiologist** [comment]
I'd call `curly_brackets` "braces" or "curlies" instead.

`_match_value` and `_assignment` are each used only once. You know what to do. Similarly, it looks like you can overwrite `target` instead of defining a new variable `subject`. "subject" is probably the better name, though.

`if guard is not None` can probably be written `if guard`.

`body=[asty.Assign(expr, targets=[return_var]`… and `elif isinstance(value, (String, Integer`… are super-long lines. Give us some newlines.

The line `print(value)` looks like a leftover debugging statement.

Add tests that statements in match bodies and `:if` clauses are executed if and only if they ought to be. For this kind of check, I often append to a list and then check the list contents afterwards.

In the test

```
(assert (is (match {}
                   {0 [1 2 {}]} 0
                   {0 [1 2 {}] 1 [[]]} 1
                   [] None)
            None))
```

I'd use something other than `None` for the last case, to distinguish from the implicit default `None`.

There's a stray newline added before `Inf = float('inf')` in `compiler.py`.

**Comment by Kodiologist** [comment]
And don't forget `(assert (is (match 1) None))`.

**Comment by Kodiologist** [comment]
I've tried to make the documentation shorter and more complete. Tell me what you think. And yeah, I think we might as well disallow patterns like `(, 1 2 3)` when the user could just write `[1 2 3]`. Goodness knows why Python allows both syntaxes if they do the same thing.

**Comment by allison-casey** [comment]
> I'd call `curly_brackets` "braces" or "curlies" instead.

changed to `braces`

> `_match_value` and `_assignment` are each used only once. You know what to do. 

inlined

> Similarly, it looks like you can overwrite `target` instead of defining a new variable `subject`. "subject" is probably the better name, though.

renamed `target` to `subject`

> `if guard is not None` can probably be written `if guard`.

done

> `body=[asty.Assign(expr, targets=[return_var]`… and `elif isinstance(value, (String, Integer`… are super-long lines. Give us some newlines.

done

> The line `print(value)` looks like a leftover debugging statement.

deleted

> Add tests that statements in match bodies and `:if` clauses are executed if and only if they ought to be. For this kind of check, I often append to a list and then check the list contents afterwards.

done

> In the test
> 
> ```
> (assert (is (match {}
>                    {0 [1 2 {}]} 0
>                    {0 [1 2 {}] 1 [[]]} 1
>                    [] None)
>             None))
> ```
> I'd use something other than `None` for the last case, to distinguish from the implicit default `None`.

changed last clause to return `2` instead of `None`

> There's a stray newline added before `Inf = float('inf')` in `compiler.py`.

deleted

> Also, it's probably not wise to test that assignments made in failed pattern matches hang around

valid, removed

> I've tried to make the documentation shorter and more complete. Tell me what you think.

I've added an example of class matching, otherwise fine by me

> And yeah, I think we might as well disallow patterns like (, 1 2 3) when the user could just write [1 2 3]. Goodness knows why Python allows both syntaxes if they do the same thing.

Not a fan. Its a deviation from both the spec and how Hy works in other areas like tuple unpacking assignment.

**Comment by Kodiologist** [comment]
> Its a deviation from… how Hy works in other areas like tuple unpacking assignment.

Good point.

**Review by Kodiologist — APPROVED** [review]
- Add a docstring for `braces`.
- I would use `(% x 2)` or something instead of `even?` because `even?` is going to be baleeted or moved to the spin-off library soon.

**Comment by allison-casey** [comment]
>     * Add a docstring for `braces`.
> 
>     * I would use `(% x 2)` or something instead of `even?` because `even?` is going to be baleeted or moved to the spin-off library soon.

Done

@scauligi you had some notes on this, how are you feeling about the current state of the PR?

File: tests/native_tests/language.hy [hunk 4]
```diff
@@ -16,7 +16,7 @@
         pytest)
 (import sys)
 
-(import [hy._compat [PY3_8]])
+(import [hy._compat [PY3_8 PY3_10]])
```
Comment (scauligi) [inline tests/native_tests/language.hy:19]: No longer needed here?

**Review by scauligi — COMMENTED** [review]
Edge case on default-case detection: we still need to add a(n unguarded) catch-all if the existing catch-all(s) are guarded:
```hy
=> (match 1
...  _ :if False 'unreachable)

match 1:
    case _ if False:
        _hy_anon_var_5 = hy.models.Symbol('unreachable')
_hy_anon_var_5

Traceback (most recent call last):
  File "stdin-a2b5526f71852fa367c6fbbc46d544b7841c58d8", line 1, in <module>
    (match 1
NameError: name '_hy_anon_var_5' is not defined
```

---

Is it a concern that the seemingly (in)correct expression
```hy
(match expr
  pat :if (cond pat) :as x  (, x pat))
```
parses (and can run without error! just with unexpected results) as the valid expression
```hy
(match expr
  pat :if (cond pat)  :as  ; the result of this match is just (Keyword "as")
  x  (, x pat)  ; a catch-all match bound to x
```

**Comment by Kodiologist** [comment]
Rather than adding a default case, it's simpler to just initialize the output variable to `None` before the `match`.

File: tests/native_tests/language.hy (hunk 4 above)
Comment (allison-casey) [reply tests/native_tests/language.hy]: removed

**Comment by allison-casey** [comment]
> Edge case on default-case detection: we still need to add a(n unguarded) catch-all if the existing catch-all(s) are guarded:
> 
> ```hy
> => (match 1
> ...  _ :if False 'unreachable)
> 
> match 1:
>     case _ if False:
>         _hy_anon_var_5 = hy.models.Symbol('unreachable')
> _hy_anon_var_5
> 
> Traceback (most recent call last):
>   File "stdin-a2b5526f71852fa367c6fbbc46d544b7841c58d8", line 1, in <module>
>     (match 1
> NameError: name '_hy_anon_var_5' is not defined
> ```

initialized the anon var to `None` and now `(match 1)` compiles to just `anon_var = None` and doesn't generate a match statement at all

> Is it a concern that the seemingly (in)correct expression
> 
> ```hy
> (match expr
>   pat :if (cond pat) :as x  (, x pat))
> ```
> 
> parses (and can run without error! just with unexpected results) as the valid expression
> 
> ```hy
> (match expr
>   pat :if (cond pat)  :as  ; the result of this match is just (Keyword "as")
>   x  (, x pat)  ; a catch-all match bound to x
> ```

made `match` raise a `HySyntaxError` if the body is `:as` and an `:if` guard exists, if you want a clause to return the keyword `:as` if now has to be quoted

```clojure
(match 1
       1 :if True :as)
Traceback (most recent call last):
  File "stdin-5322fe22a66153576430f80f7d8b627c931eda28", line 1, in <module>
    (match 1 1 :if True :as )
  File "<stdin>", line 1
    (match 1 1 :if True :as )
                        ^
hy.errors.HySyntaxError: :as clause cannot come after :if guard
```

```clojure
(match 1
       1 :if True ':as)
:as
```

**Comment by Kodiologist** [comment]
> initialized the anon var to `None` and now `(match 1)` compiles to just `anon_var = None` and doesn't generate a match statement at all

You still need to evaluate the subject, though, in case it has side-effects.

**Comment by allison-casey** [comment]
its still compiled and added to the Result
```python
ret = Result() + subject
ret += asty.Assign(expr, targets=[return_var], value=asty.Constant(expr, value=None))
if not match_cases:
    return ret + returnable
```

**Comment by Kodiologist** [comment]
You need an `expr_as_stmt`, though:

```
$ hy --spy
import hy
hy 1.0a1+156.g8f4476f8 using CPython(default) 3.10.0b2 on Linux
=> (match (f))
_hy_anon_var_1 = None
_hy_anon_var_1
```

**Comment by allison-casey** [comment]
ahhh got it. Done

**Review by scauligi — APPROVED** [review]
heccin' _DOPE_
